### PR TITLE
Add impl trait declarations to SMIR

### PR DESCRIPTION
**Pull request by spastorino**
r? @oli-obk

### CODE REVIEW

File: compiler/rustc_smir/src/rustc_smir/mod.rs [hunk 1]
```diff
@@ -56,6 +56,20 @@ impl<'tcx> Context for Tables<'tcx> {
...
+        self.tcx
+            .trait_impls_in_crate(LOCAL_CRATE)
+            .iter()
+            .map(|impl_def_id| self.impl_def(*impl_def_id))
+            .collect()
+    }
+
+    fn trait_impl(&mut self, impl_def: &stable_mir::ty::ImplDef) -> stable_mir::ty::TraitRef {
+        let def_id = self.impl_trait_def_id(impl_def);
+        let impl_trait_def = self.tcx.impl_trait_ref(def_id).unwrap().skip_binder();
```
Comment (spastorino) [inline compiler/rustc_smir/src/rustc_smir/mod.rs:69]: @oli-obk given that this is an `EarlyBinder` I think we can just `skip_binder` here to simplify things.

Comment (oli-obk) [reply compiler/rustc_smir/src/rustc_smir/mod.rs:69]: I believe what we want is `instantiate_identity()`, which signals that we are not ignoring arguments, but instead expect to be using it in the parameter environment of the parent.

That said, we should probably start preserving early binders

File: compiler/rustc_smir/src/rustc_smir/mod.rs [hunk 2]
```diff
@@ -56,6 +56,20 @@ impl<'tcx> Context for Tables<'tcx> {
         trait_def.stable(self)
     }
 
+    fn all_trait_impls(&mut self) -> stable_mir::ImplTraitDecls {
+        self.tcx
+            .trait_impls_in_crate(LOCAL_CRATE)
+            .iter()
+            .map(|impl_def_id| self.impl_def(*impl_def_id))
```
Comment (oli-obk) [inline compiler/rustc_smir/src/rustc_smir/mod.rs:63]: Please make this lazy and only return a new kind of `ImplDefId`

File: compiler/rustc_smir/src/stable_mir/mod.rs [hunk 3]
```diff
@@ -89,6 +92,8 @@ pub trait Context {
     fn mir_body(&mut self, item: &CrateItem) -> mir::Body;
     fn all_trait_decls(&mut self) -> TraitDecls;
     fn trait_decl(&mut self, trait_def: &TraitDef) -> TraitDecl;
+    fn all_trait_impls(&mut self) -> ImplTraitDecls;
+    fn trait_impl(&mut self, trait_impl: &ImplDef) -> TraitRef;
```
Comment (oli-obk) [inline compiler/rustc_smir/src/stable_mir/mod.rs:96]: Add a method to `ImplDef` that calls this

File: compiler/rustc_smir/src/stable_mir/ty.rs [hunk 4]
```diff
@@ -285,3 +299,10 @@ pub struct TraitDecl {
     pub implement_via_object: bool,
     pub deny_explicit_impl: bool,
 }
+
+pub type ImplTrait = EarlyBinder<TraitRef>;
+
+pub struct TraitRef {
+    pub def_id: ImplDef,
```
Comment (oli-obk) [inline compiler/rustc_smir/src/stable_mir/ty.rs:306]: This should be a `TraitDef`, as the `DefId` refers to the trait, not the impl

Comment (spastorino) [reply compiler/rustc_smir/src/stable_mir/ty.rs:306]: Uff, right ... my bad.

**Pull request merged**